README.md
# Scyland3D: Processing 3D landmarks

``Scyland3D`` is a Python tool for converting 3D raw landmark and semilandmark coordinates exported from landmark acquisition software (e.g. [Landmark Editor 3.6 from IDAV, UC Davis](http://www.idav.ucdavis.edu/research/EvoMorph)) to a csv format best suited for geometric morphometric analyses.
Processing schemes to mirror and reorder these points are provided to address further symmetry issues.

By:
- Fidji Berio ([GitHub](https://github.com/fberio)) from [ISEM](http://www.isem.univ-montp2.fr/en/) and [IGFL](http://igfl.ens-lyon.fr/igfl/annuaire/berio-fidji),
- Yann Bayle ([Website](http://yannbayle.fr/english/index.php), [GitHub](https://github.com/ybayle)) from [LaBRI](http://www.labri.fr/), [Univ. Bordeaux](https://www.u-bordeaux.fr/), [CNRS](http://www.cnrs.fr/), and [SCRIME](https://scrime.u-bordeaux.fr/).

## Installation

Install and update using pip:

`pip install -U Scyland3D`

Dependencies:

- [Python](https://www.python.org/) 2 or 3. **Scyland3D** has been tested with Python 2.7.16 and 3.6.3 and **should** work with older versions of Python.
- [Numpy](https://www.numpy.org/) >= 1.13.3. **Scyland3D** has been tested with Numpy 1.13.3 and 1.16.2 and **should** work with older versions of Numpy.

If you encounter an error with other versions, please [submit an issue](https://github.com/ybayle/Scyland3D/issues/new).

## Use-case example

To quickly process landmarks files, you can give **Scyland3D** the path to the folder containing the `.pts` files (format defined on page 37 in http://www.idav.ucdavis.edu/research/projects/EvoMorph/supplement/LandmarkDoc_v3_b6.pdf), as shown in the example files provided:

```
from Scyland3D import Scyland3D
Scyland3D.pts2csv("example/")
```

This will create a file named `landmarks.csv` that you can use for statistical analyses in any language/software you want.
The example files provided come from a subset of our dataset for studying teeth of the small-spotted catshark *Scyliorhinus canicula*.
We provide 7 landmark and 31 semilandmark coordinates for 2 upper teeth and for 3 lower ones, stored in two folders that describe the specimens they belong to.
A landmark is a point set up by hand as opposed to a semilandmark that is interpolated by the computer between two landmarks by following the curvature of the studied form.

Below, the image A is the representation of 7 landmarks for an input right shark tooth and the image B is the output produced by Scyland3D when using the options to mirror and reorder the landmarks.
This process can be used to process right teeth in order to superimpose and compare them with left teeth in order to compare analogous elements.
More information can be found on the [related article](https://github.com/openjournals/joss-papers/blob/joss.01153/joss.01153/10.21105.joss.01153.pdf).

![Landmarks on a tooth](https://raw.githubusercontent.com/ybayle/Scyland3D/master/paper/figure1.png)

## API documentation

The function `pts2csv()` is the core of **Scyland3D** and is also called when using the script from the terminal with the following arguments:

- *indir* (required)
    - A string containing the input directory where the files are stored. File names can specify the feature modalities separated by `_` (e.g. speciesA_ageX_sex1.pts).
    - Example: `Scyland3D.pts2csv(indir="path/")`
    - Command line: `python -m Scyland3D.Scyland3D -i "path/"`
- *outdir* (optional)
    - A string containing the output directory where the files will be generated.
    - Default: Use the current folder.
    - Example: `Scyland3D.pts2csv(indir="path/", outdir="outpath/")`
    - Command line: `python -m Scyland3D.Scyland3D -i "path/" -o "path/"`
- *order* (optional)
    - A list of integer indicating the new order to apply to the landmarks. The order_factor argument must also be supplied.
    - Default: None
    - Example: `Scyland3D.pts2csv(indir="path/", order_factor="upper", order=[36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37])`
    - Command line: `python -m Scyland3D.Scyland3D -i "path/" -f "upper" -r "36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37"`
- *order_factor* (optional)
    - A string containing the keyword for items that need their landmarks and semilandmarks to be reordered. The order argument must also be supplied. **Warning:** no checks are made on the factor, so be sure to avoid any typo.
    - Default: None
    - Example: if the dataset contains lower and upper teeth where only the upper teeth need to be reordered, one can use: `Scyland3D.pts2csv(indir="path/", order_factor="upper", order=[36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37])`
    - Command line: `python -m Scyland3D.Scyland3D -i "path/" -f "upper" -r "36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37"`. This is only valid if the keyword `upper` is present in the file name.
- *feature_names* (optional)
    - A list of string containing the feature names.
    - Default: None
    - Example: `Scyland3D.pts2csv(indir="path/", feature_names=["identifier", "species", "location", "length", "sex", "stage", "jaw", "position", "generation"])`
    - Command line: `python -m Scyland3D.Scyland3D -i "path/" -n "identifier,species,location,length,sex,stage,jaw,position,generation"`
- *verbose* (optional)
    - A boolean indicating if information should be printed on the console by the process.
    - Default: False
    - Example: `Scyland3D.pts2csv(indir="path/", verbose=True)` displays information to the console.
    - Command line: `python -m Scyland3D.Scyland3D -i "path/" -v` displays information to the console.
- *mirror_factor* (optional)
    - A string containing the keyword for items to be mirrored in the 3D space. **Warning:** no checks are made on the factor, so be sure to avoid any typo.
    - Default: None
    - Example: `Scyland3D.pts2csv(indir="path/", mirror_factor="upper")`
    - Command line: `python -m Scyland3D.Scyland3D -i "path/" -m "upper"` will mirror in 3D the landmarks stored in `.pts` files with the keyword upper in their filename. Then, it will export to a `.csv` file those mirrored landmarks and also the unprocessed landmarks from `.pts` files that do not contain the keyword `upper` in their filename. The figure above depicts the mirroring.

## How To Contribute

Contributions are welcome!
Please refer to the [CONTRIBUTING.md](CONTRIBUTING.md) file.

## License

**Scyland3D** is licensed under the MIT License as described in the [license file](LICENSE). Please use the following citation:

```
@article{Berio2019,
    author  = {Berio, Fidji and Bayle, Yann},
    title   = {{Scyland3D: Processing 3D landmarks}},
    journal = {{The Journal of Open Source Software}},
    page    = {1153--1155},
    volume  = {4},
    issue   = {33},
    doi     = {10.21105/joss.01153},
    note    = {Review pending}
}
```

## Acknowledgements

- We acknowledge the contribution of SFR Biosciences (UMS3444/CNRS, US8/Inserm, ENS de Lyon, UCBL) facilities: AniRA-ImmOs and Mathilde Bouchet for her help with X-ray microtomography.
- We are thankful to @patrikhuber and @r-barnes, the two reviewers from [JOSS](https://github.com/openjournals/joss-reviews) that helped improved a lot the code and the documentation.


CHANGELOG.md
# CHANGELOG

## 1.2.0 2021-04-26

Add 3D to 2D option.

## 1.1.0 2020-02-04

Generated Github version.

## 1.0.20 2020-02-04

Fix typos in article and generate doi.

## 1.0.19 2019-11-03

Fix #12 about invalid parameters when using the command line.
Fix #13 about incomplete readme details.
Improve the documentation.

## 1.0.18 2019-10-20

Add details in ReadMe about the context of the included image #5
Refactor code's comments #8
Fix file names order consistency on all OS #11

## 1.0.17 2019-10-19

Fix ReadMe for package usage #10
Fix test.sh #10
Move from getopt to argparse

## 1.0.7 2019-06-26

Add test files
Format with black

## 1.0.6 2019-06-25

Add no regression tests

## 1.0.5 2019-06-24

Add command line support

## 1.0.4 2019-06-23

Python refactoring

## 1.0.3 2019-06-12

Python refactoring and added comment in the code #3 #7 #8 #9

## 1.0.2 2019-03-30

Add change log
Add arguments parsing #1

## 1.0.1 2019-03-30

Enhance community guidelines #2
Add python 2 and 3 compatibility message #4
Use python core csv module #9 

## 1.0.0 2019-01-03

First release


test.sh
echo 'Testing call from python...'
python -c 'import Scyland3D; Scyland3D.test_no_regression()'
echo 'Testing call from the command line...'
python Scyland3D.py -i "example/"
python Scyland3D.py -i "example/" -f "upper" -r "36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37"
python Scyland3D.py -i "example/" -m "upper" -n "identifier,species,location,length,sex,stage,jaw,position,generation"
python -c 'import Scyland3D; Scyland3D._validation_against_ref()'


Scyland3D/test/test_project3Dto2D.py
# -*- coding: utf-8 -*-
#
# Authors   Fidji Berio and Yann Bayle
# E-mails   fidji.berio@ens-lyon.fr and bayle.yann@live.fr
# License   MIT
# Created   26/04/2021
#

import os
import sys
sys.path.insert(0, os.path.join(__file__, "..", ".."))
import Scyland3D
from mpl_toolkits import mplot3d
import numpy as np
import matplotlib.pyplot as plt

class TestsProject3Dto2D:
    def test_exportToDelete(self):
        dirn = os.path.join(os.path.dirname(__file__), "")
        filen = dirn + "3d_hatchlings.txt"
        new_str = ""
        with open(filen, "r") as filep:
            new_str += next(filep)
            for line in filep:
                points3D = line
                points3D = points3D.split(",")[1:3*38+1]
                xs = []
                ys = []
                zs = []
                for i in np.arange(len(points3D), step=3):
                    xs.append(float(points3D[i]))
                    ys.append(float(points3D[i+1]))
                    zs.append(float(points3D[i+2]))
                data = []
                for val_x, val_y, val_z in zip(xs, ys, zs):
                    data.append([val_x, val_y, val_z])

                newData = Scyland3D._reverse_z(data)

                new_str += line.split(",")[0] + ","
                for xyz in newData:
                    new_str += str(xyz[0]) + "," + str(xyz[1]) + "," + str(xyz[2]) + ","
                new_str += ",".join(line.split(",")[3*38+1:])
        with open(dirn + "hatchlings_2D_plane_in_3D.txt", "w") as filep:
            filep.write(new_str)

        assert False

    def test_project3Dto2D(self):
        
        def displayPoints(xs, ys, zs):
            plt.figure()
            ax = plt.axes(projection='3d')
            # ax.plot3D(xs, ys, zs)
            ax.scatter(xs, ys, zs)
            plt.show()

        dirn = os.path.join(os.path.dirname(__file__), "")
        filen = dirn + "3d_hatchlings.txt"
        with open(filen, "r") as filep:
            next(filep)
            points3D = next(filep)
        points3D = points3D.split(",")[1:3*38+1]
        xs = []
        ys = []
        zs = []
        for i in np.arange(len(points3D), step=3):
            xs.append(float(points3D[i]))
            ys.append(float(points3D[i+1]))
            zs.append(float(points3D[i+2]))

        displayPoints(xs, ys, zs)

        data = []
        for val_x, val_y, val_z in zip(xs, ys, zs):
            data.append([val_x, val_y, val_z])

        newData = Scyland3D._reverse_z(data)
        new_xs = []
        new_ys = []
        new_zs = []
        for xyz in newData:
            new_xs.append(xyz[0])
            new_ys.append(xyz[1])
            new_zs.append(xyz[2])
        displayPoints(new_xs, new_ys, new_zs)

        # points2D = Scyland3D.project3Dto2D(points3D)

        plt.figure()
        plt.plot(new_xs, new_ys, "o")
        plt.show()

        assert False


Scyland3D/Scyland3D.py
# -*- coding: utf-8 -*-
#
# Authors   Fidji Berio and Yann Bayle
# E-mails   fidji.berio@ens-lyon.fr and bayle.yann@live.fr
# License   MIT
# Created   15/02/2018
#

import os
import csv
import sys
import argparse
import numpy as np


def _export2csv(data, nb_landmark, outdir=None, feature_names=None, modif=""):
    """export2csv
    Export data to a CSV named indir + "../landmarks" + modif + ".csv".

    Args:
        data (array): The landmarks to export.
        nb_landmark (int): The number of landmarks.
        outdir (str): The name of the directory where the output files will be stored. If not
            provided the current working directory will be used.
        feature_names (str or array of str): String describing each feature names separated with a
            comma or array of string (e.g. "age,sex,size" if supplied from the command line or
            ["age", "sex", "size"] if supplied in python script).
        modif (str): The name of the modification applied to the data (e.g. none, reversed, and/or
            reordered).
    """
    # Generate the header of the csv and store it in fieldnames
    fieldnames = ["ID"]
    assert nb_landmark > 0, "The number of landmarks must be positive."
    # First add the numbered coordinates (e.g. x1,y1,z1,...,xN,yN,zN)
    for numb in range(1, nb_landmark + 1):
        for axe in ["x", "y", "z"]:
            fieldnames.append(axe + str(numb))
    # Gather the number of features, i.e. observable variables, that corresponds to the length
    # (i.e. number of columns) of one row of the data, minus the number of 3D coordinates,
    # minus one that corresponds to the ID. Example:
    # ID,x1,y1,z1,x2,y2,z2,...xN,yN,zN,feature1,feature2,...,featureM
    # 200118G,25.6,11.6,23.9,26.5,23.5,14.9,...,25.5,11.5,23.8,mediterranean,female,...,mature
    nb_feature = len(data[0]) - nb_landmark * 3 - 1
    # Use generic feature names for the header if no feature_names has been provided by the user
    if feature_names is not None:
        if isinstance(feature_names, str):
            feature_names = feature_names.split(",")
        assert nb_feature == len(feature_names), (
            "The number of feature names provided ("
            + str(len(feature_names))
            + ") does not match the number of features detected in the files ("
            + str(nb_feature)
            + ")."
        )
        for feature_name in feature_names:
            fieldnames.append(feature_name)
    else:
        for numb in range(1, nb_feature + 1):
            fieldnames.append("Feature" + str(numb))
    # Use the current folder if no output directory has been supplied by the user
    if outdir is None:
        outdir = "./"
    output_filename = os.path.join(os.path.abspath(outdir), "landmarks" + modif + ".csv")
    # Export the header and the data
    # newline='' is use to fix #11. On Windows the csv package add an extra '\r' that is not
    # observed on Unix systems.
    # Additionally, this fix only works on python 3 as python 2 open's does not support the
    # `newline` argument.
    if sys.version_info[0] == 2:
        access = 'wb'
        kwargs = {}
    else:
        access = 'wt'
        kwargs = {'newline':''}
    with open(output_filename, access, **kwargs) as out_file:
        csv.DictWriter(out_file, fieldnames=fieldnames).writeheader()
        csv.writer(out_file).writerows(data)
    print("File successfully generated: " + output_filename)


def _list_pts(indir):
    """list_pts
    Recursively lists all .pts file from indir.

    Args:
        indir (str): The name of the directory where the .pts files are stored.

    Returns:
        An array of .pts file names to process. 
    """
    assert os.path.exists(indir) and os.path.isdir(indir), indir + " not found."
    filenames = [
        os.path.join(dp, f)
        for dp, dn, fn in os.walk(os.path.expanduser(indir))
        for f in fn
    ]
    filenames = [filen for filen in filenames if ".pts" in filen]
    # Fix #11 by sorting filenames to guarantee consistency between OS.
    filenames.sort()
    assert filenames, "There are no .pts files in " + indir
    return filenames


def _remove_duplicates(data):
    """remove_duplicates
    Remove the duplicates landmarks and semi-landmarks.

    Args:
        data (array): Contains the landmarks.

    Returns:
        An array of landmarks where duplicates were removed.
    """
    data = data[7:]
    # Remove easy duplicates in the coordinates where the string are an exact match
    data = sorted(set(data), key=data.index)
    # Remove the other duplicates because some landmarks and semilandmarks are sometimes exported twice.
    coord = []
    # Convert the coordinates from str to float
    for item in data:
        coord.append([float(xyz) for xyz in item.split(",")])
    index_to_remove = []
    # Check which landmarks have similar coordinates up to epsilon
    epsilon = 1e-6 # needs to be between 4e-5 and 4e-7
    for i in range(len(coord) - 1):
        for j in range(len(coord) - i - 1):
            if (abs(coord[i][0] - coord[i + j + 1][0]) < (epsilon * coord[i][0]) and abs(coord[i][1] - coord[i + j + 1][1]) < (epsilon * coord[i][1]) and abs(coord[i][2] - coord[i + j + 1][2]) < (epsilon * coord[i][2])):
                index_to_remove.append(i + j + 1)
    # Return only landmarks that have not been detected as duplicates 
    new_data = []
    for i, item in enumerate(coord):
        if i not in index_to_remove:
            new_data.append(item)
    return new_data


def _reverse_z(data):
    """reverse_z
    Find the relative z-axis for a given set of point by computing a plane that minimizes the
    difference between it and the given points.
    Then reverses the set of points relatively to this new plane, so no user specific plane has to
    be supplied.
    Inspired from https://www.youtube.com/watch?v=86lwcXeZoiA and the suggestion from
    https://stackoverflow.com/questions/8954326/how-to-calculate-the-mirror-point-along-a-line#8954454
    is not applicable here as it is not valid in 3D and does not take into account relative axes. 

    Args:
        data (array): Contains the landmarks.

    Returns:
        An array of landmarks inverted along the z-axis.
    """
    matrix_xy = []
    matrix_z = []
    # Split the coordinates in an xy plane and a z axis
    for xyz in data:
        x = xyz[0]
        y = xyz[1]
        z = xyz[2]
        matrix_xy.append([x, y, 1])
        matrix_z.append(z)
    # Gather the coefficients for a linear fitting
    b = np.matrix(matrix_z).T
    A = np.matrix(matrix_xy)
    # Compute the linear fit for the plane going through all landmarks points
    fit = (A.T * A).I * A.T * b

    # Gather the parameters for the plane (A,B,C,D) defined by A*x + B*y + C*z + D
    A = fit[0]
    B = fit[1]
    # Compute the inversion coefficients for the mirroring
    C = -1.0
    D = -fit[2]
    # C = 0
    # D = 0
    data_mirror = []
    for xyz in data:
        x = xyz[0]
        y = xyz[1]
        z = xyz[2]
        # Compute the symmetry point that is the intersection between the plane (A,B,C,D) and a
        # perpendicular line to that plane that go through the point of interest defined by (x,y,z).
        t = float((D - A * x - B * y - C * z) / (A * A + B * B + C * C))
        # Compute the resulting value for the mirrored points where the new coordinate is equal to
        # twice the distance between the old coordinate and the symmetry point.
        xp = float(x + A * t)
        yp = float(y + B * t)
        zp = float(z + C * t)
        # xp = float(x + 2.0 * A * t)
        # yp = float(y + 2.0 * B * t)
        # zp = float(z + 2.0 * C * t)

        # https://stackoverflow.com/questions/23472048/projecting-3d-points-to-2d-plane ?
        # n = np.array([A, B, C])
        # x_axis = np.array([1, 0, 0])
        # x_axis = x_axis - np.dot(x_axis, n) * n
        # newValue = np.sqrt((x_axis**2).sum())
        # print(newValue)
        # print(x_axis)
        # print(x_axis / newValue)
        # x_axis = x_axis / newValue
        # # x_axis /= np.sqrt((x_axis**2).sum())
        # y_axis = np.cross(n, x_axis)
        # myPoint = np.array(xp, yp, zp)
        # xp = np.dot(myPoint, x_axis)
        # yp = np.dot(myPoint, y_axis)
        # zp = 0


        data_mirror.append([xp, yp, zp])
    return data_mirror


def _get_path(filen):
    """_get_path
    Return the absolute path of the provided file name from the absolute resources directory.

    Args:
        filen (str): The name of the file in the relative path of the package.

    Returns:
        An absolute path to the custom resources provided with the package.
    """
    return os.path.join(os.path.abspath(os.path.dirname(__file__)), filen)


def pts2csv(
    indir=None,
    outdir=None,
    mirror_factor=None,
    order=None,
    order_factor=None,
    feature_names=None,
    verbose=False,
):
    """pts2csv
    Convert .pts files from indir to a single .csv file

    Args:
        indir (str): The name of the directory where the output files will be stored.
        mirror_factor (str): The name of the factor to use for mirroring the landmarks on the z-axis.
        order (str or array of int): A string (if supplied by the command line) or an array of int
        (if called from another python script) indicating how to reorder the landmarks (e.g. "1,3,2"
        if supplied from the command line or [1, 3, 2] if supplied by another python script).
        order_factor (str): The name of the factor to use for reordering the landmarks.
        feature_names (str or array of str): String describing each feature names separated with a
        comma or array of string (e.g. "age,sex,size" if supplied from the command line or 
        ["age", "sex", "size"] if supplied in python script).
        verbose (bool): Whether to output details during the process.
    """
    assert os.path.exists(indir) and os.path.isdir(indir), indir + " not found."
    assert (order is None and order_factor is None) or (
        order is not None
        and order_factor is not None
        and (isinstance(order_factor, str) or isinstance(order_factor, list))
    ), "Must supply order and order_factor."
    if indir[-1] != os.sep:
        indir += os.sep

    nb_feature = 0
    nb_landmark = 0
    list_pts_files = _list_pts(indir)
    data2write = []
    order_factor_found_at_least_in_one_file = False
    mirror_applied_at_least_in_one_file = False
    # For each .pts file
    for index, filen in enumerate(list_pts_files):
        if verbose:
            print(str(index + 1) + "/" + str(len(list_pts_files)) + " " + filen)
        # Count the number of feature in that file
        nb_detected_feature = filen[filen.find(os.sep) + 1 :].count("_") + filen.count(
            os.sep
        )
        if nb_feature == 0:
            # Use the number of features from the first .pts file as reference
            nb_feature = nb_detected_feature
        else:
            # Check that the number of feature between all .pts files are consistent
            assert nb_feature == nb_detected_feature, (
                "The name of the file or of the directory does not seems to be consistent because for "
                + filen
                + " because there are "
                + str(nb_detected_feature)
                + " detected features while "
                + str(nb_feature)
                + " were expected."
            )
        data = []
        # Gather semi-landmarks (S) and landmarks (C)
        with open(filen, "r") as filep:
            for line in filep:
                if line[0] in ["S", "C"]:
                    data.append(",".join(line[:-1].split("  ")[1:]))
        # Remove duplicates landmarks generated that are generally present in the .pts files
        data = _remove_duplicates(data)
        if nb_landmark == 0:
            nb_landmark = len(data)
        assert len(data) == nb_landmark, (
            "Some landmarks may not be correctly superimposed for "
            + filen
            + " because there are "
            + str(len(data))
            + " landmarks detected instead of "
            + str(nb_landmark)
        )
        # Apply a z-axis mirror to the landmarks to study left-right differences in the given species
        if (
            mirror_factor is not None
            and isinstance(mirror_factor, str)
            and mirror_factor in filen
        ):
            data = _reverse_z(data)
            mirror_applied_at_least_in_one_file = True
        # Reorder the landmarks as specified by the order argument only for the files containing the
        # order_factor string.
        if order_factor is not None and order_factor in filen:
            order_factor_found_at_least_in_one_file = True
            if isinstance(order, str):
                order = [int(val) for val in order.split(",")]
            data = np.array(data)
            data = data[order]
            data = data.tolist()
        # Add a new row to the data
        # This first column contains an unique ID for each sample that corresponds to the file path
        # and name.
        # .replace("//", "/") replaces "//" by "/" that occurs on Windows. This is used to enable
        # comparisons with reference files in tests for all OS.
        row = ["/".join(filen.split(os.sep)[-2:]).replace("//", "/")]
        data = np.array(data)
        for val in data.reshape(data.shape[0] * data.shape[1]):
            row.append(val)
        filen = filen.split(".")[-2]
        for feature in (
            filen[filen.find(os.sep) + 1 :]
            .replace("_", ",")
            .replace("-", ".")
            .split(os.sep)[-1]
            .split(",")
        ):
            row.append(feature)
        data2write.append(row)
    if order_factor is not None:
        assert order_factor_found_at_least_in_one_file, (
            "The order_factor ("
            + order_factor
            + ") provided has not been found in any file names."
        )
    if mirror_factor is not None:
        assert mirror_applied_at_least_in_one_file, (
            "The mirror_factor ("
            + mirror_factor
            + ") provided has not been found in any file names."
        )
    modif = ""
    if order_factor_found_at_least_in_one_file:
        modif += "_reordered"
    if mirror_applied_at_least_in_one_file:
        modif += "_reversed"
    _export2csv(
        outdir=outdir,
        data=data2write,
        nb_landmark=nb_landmark,
        feature_names=feature_names,
        modif=modif,
    )


def _same_file(filen1, filen2):
    """_same_file
    Return True if filen1 and filen2 contains the same data

    Args:
        filen1 (str): The path and name of the first filename
        filen2 (str): The path and name of the second filename

    Returns:
        A bool indicating if the two files contains the same data 
    """
    assert filen1 != filen2, "File names must be different."
    with open(filen1, "r") as filep_ref:
        with open(filen2, "r") as filep_test:
            for line_ref, line_test in zip(filep_ref, filep_test):
                assert line_ref == line_test, "Invalid line generated:\n" + line_test + "\nthat is different from the reference file:\n" + line_ref
    return True


def _same_file_up_to_epsilon(filen1, filen2, eps=1e-9):
    """_same_file_up_to_epsilon
    Return True if filen1 and filen2 contains the same float data up to epsilon

    Args:
        filen1 (str): The path and name of the first filename
        filen2 (str): The path and name of the second filename
        eps (float): The maximum tolerance for asserting that two floating point
                     numbers are different

    Returns:
        A bool indicating if the two files contains the same data 
    """
    assert filen1 != filen2, "File names must be different."
    with open(filen1, "r") as filep_ref:
        with open(filen2, "r") as filep_test:
            line_ref = next(filep_ref)
            line_test = next(filep_test)
            assert line_ref == line_test, "Invalid line generated for " + filen1 + ":\n" + line_test + "\nthat is different from the reference file " + filen2 + ":\n" + line_ref
            for line_ref, line_test in zip(filep_ref, filep_test):
                # Checks that the 38 3D landmarks generated are equal up to epsilon to the reference
                for val_ref, val_test in zip(line_test.split(",")[1:38 * 3 + 1], line_test.split(",")[1:38 * 3 + 1]):
                    assert abs(float(val_ref) - float(val_test)) < eps, "Invalid value detected for " + filen1 + ":\n" + line_test + "\nthat is different from the reference file " + filen2 + ":\n" + line_ref
    return True


def _validation_against_ref():
    """_validation_against_ref
    
    Tests whether the generated file corresponds to the reference file to check for regressions

    Args:
        None
    Returns:
        None
    """
    print("Testing default behavior...")
    dirn = _get_path("./")
    assert _same_file(
        dirn + "landmarks.csv", dirn + "test/landmarks_ref.csv"
    ), "Generated file does not match the default reference."
    print("Ok!")
    print("Testing reordering...")
    assert _same_file(
        dirn + "landmarks_reordered.csv", dirn + "test/landmarks_reordered_ref.csv"
    ), "Generated file does not match the reference for the reordering."
    print("Ok!")
    print("Testing reversing...")
    # The reversing applied mathematical operations on the data so numerical
    # differences is observed between OS. A comparison up to epsilon is thus
    # made.
    assert _same_file_up_to_epsilon(
        dirn + "landmarks_reversed.csv", dirn + "test/landmarks_reversed_ref.csv"
    ), "Generated file does not match the reference for the reversing."
    print("Ok!")


def project3Dto2D(points3D):
    """
    project3Dto2D

    Projects the set of 3D points in 2D and return the new points. 

    Args:
        points3D ([list]): Input points in 3D.
    """
    points2D = [2]
    
    return points2D


def test_no_regression():
    """test_no_regression

    Verify that no regression are found between the new generated files and the reference files

    Args:
        None
    Returns:
        None
    """
    indir = _get_path("example/")
    verbose = False
    order_factor = "upper"
    outdir = _get_path("./")
    order = [
        36,
        35,
        34,
        33,
        32,
        31,
        30,
        29,
        28,
        27,
        26,
        25,
        24,
        23,
        22,
        21,
        20,
        19,
        18,
        17,
        16,
        15,
        14,
        13,
        12,
        11,
        10,
        9,
        8,
        7,
        6,
        5,
        4,
        3,
        2,
        1,
        0,
        37,
    ]
    mirror_factor = "upper"
    feature_names = [
        "identifier",
        "species",
        "location",
        "length",
        "sex",
        "stage",
        "jaw",
        "position",
        "generation",
    ]
    pts2csv(
        indir=indir,
        outdir=outdir,
        mirror_factor=None,
        order=None,
        order_factor=None,
        feature_names=None,
        verbose=verbose,
    )
    pts2csv(
        indir=indir,
        outdir=outdir,
        mirror_factor=None,
        order=order,
        order_factor=order_factor,
        feature_names=None,
        verbose=verbose,
    )
    pts2csv(
        indir=indir,
        outdir=outdir,
        mirror_factor=mirror_factor,
        order=None,
        order_factor=None,
        feature_names=feature_names,
        verbose=verbose,
    )
    _validation_against_ref()


if __name__ == "__main__":
    """main entry point
    Parse arguments and call the function to convert multiple .pts files to a single csv file
    """
    parser = argparse.ArgumentParser(description='Scyland3D: Processing 3D landmarks.')
    parser.add_argument('-i', '--indir', type=str, help='the input directory containing the landmarks to process')
    parser.add_argument('-o', '--outdir', help='the output directory containing the landmarks to process')
    parser.add_argument('-m', '--mirror_factor', help='the factor to be used when mirroring the corresponding landmarks')
    parser.add_argument('-r', '--order', help='the new order to apply to the landmarks')
    parser.add_argument('-f', '--order_factor',help='the factor to be used when reordering the corresponding landmarks')
    parser.add_argument('-n', '--feature_names', help='the feature names to use in the header in the output file')
    parser.add_argument('-v', '--verbose', default=False, action="store_true", help='whether to output the processing steps')
    if len(sys.argv) == 1:
        parser.print_help(sys.stderr)
        sys.exit(1)
    args = parser.parse_args()
    pts2csv(args.indir, args.outdir, args.mirror_factor, args.order, args.order_factor, args.feature_names, args.verbose)
